### ISSUE TITLE
Regular ingest of METS files from storage service

### ISSUE THREAD

**Issue by jtweed** [wellcometrust/platform#3921]
## Notes
- [ ] Get a bag looks like from the storage service
- [ ] Extract and parse METS from a bag

## Questions
- *Is the storage service the source data?*
   Or could we listen to Goobi potentially?

- *Do we need an adapter?* 
   Yes, in order to provide standard reindexes with data model changes an index of records is required (though it need not contain a full summary of the METS XML).

- *How do we do the first __full__ ingest of the METS files?*
  ???

- *Are we going to have to deal with partial updates?*
   Given that a bag might change, but the METS file might not.

- *What is the identifier for a METS file?*
   @alicefuzier and @jamesgorrie spoke about namespacing IDs with their source e.g. `Goobi/b123456` so we don't get stuck i na world of `B` numbers being replicated across systems.

**Comment by alicefuzier** [comment]
After talking a bit with everyone this is what we came up with

- Adapter
  - receive and store message from storage service with bag url
  - send bag url to transformer
(Adapter is needed to be able to replay stuff through the pipeline as often as we want. It's not the storage service responsibility to offer that functionality)

- Transformer
  - Receive a message with the bag url
  - Call storage service and get the bag (eg `{{gateway_endpoint}}/bags/digitised/b30246039`)
  - Get the METS file from s3
  - Extract `mods:recordIdentifier` to get the bnumber and `mods:accessCondition type="dz"` to get the license
   - Create a `UnidentifiedInvisibleWork` with 
     - namespaced bnumber as `sourceIdentifier` and
     - a digital item with license
     - a `mergedCandidate` referencing the sierra work

- Recoder: no change 🤞 

- Matcher
  - remove condition to ignore `UnidentifiedInvisibleWork`
 
- Merger:
  - remove condition to skip rules for `UnidentifiedInvisibleWork`
  - Add rule to add digital item to sierra work

**Comment by tomcrane** [comment]
(a note from outside observation)
For digitised works, the AccessConditon is not a property of the work but of parts of the structure of the work. Sometimes this is the root structure, therefore it effectively applies to every file (page, video etc) in the work. But sometimes, one or two pages out of 100 in an archive item are restricted, but the rest just require clickthrough.

**Comment by alicefuzier** [comment]
Ah, that's interesting.. In the current model, a Work can have a list of `Item`s and each item has a list of `Location`s. A `Location` can have `License`. 
My understanding is that for METS we would create one `Item` with one iiif-presentation `Location`. 
Should we create multiple items and/or multiple locations with different licenses if there are multiple images?

**Comment by tomcrane** [comment]
What I don't know without checking is whether the dz license code ever changes at different parts of the structure. It _could_, and the AccessCondition `status` certainly does. I'm in tomorrow btw.

**Comment by warrd** [comment]
I have a gist [here](https://gist.github.com/warrd/0b1ee98e6e9bce47e0b236c5358a610f) about an issue implementing this feature regarding the `Work` model in the scala code.

Me and @alicefuzier think option 3 is by far the best option, but would be good to hear other opinions!

**Comment by jtweed** [comment]
Re the conversation above above different licensing within a work, I'm not too concerned about that at this stage.

We'll be using the data for search filtering, not generating the reading experience, which we remain using the existing IIIF APIs.

**Comment by jamesgorrie** [comment]
I am assuming though we will create __ONE__ CC license per item?

**Comment by tomcrane** [comment]
btw this may be of interest:
https://iiif.io/api/presentation/3.0/change-log/#125-rename-license-to-rights

**Comment by tomcrane** [comment]
So to summarise

In your digital objects currently, one license/rights is _asserted_ at the object level, on the Manifest.
And access conditions are _enforced_ on individual resources (images/video etc) at the web resource level, individually for the component resources of the compound digital object; as they must be, because you interact with resources one discrete HTTP request at a time.

**Comment by alicefuzier** [comment]
I'm having second thoughts about the design that I described above. I was hoping to talk about it today but since I'm not in the office, I’m going to write my doubts here.
This is the bit of the design that I'm not convinced about:

> - Adapter (Adapter is needed to be able to replay stuff through the pipeline as often as we want. It's not the storage service responsibility to offer that functionality)
>   - receive and store message from storage service with bag url 
>   - send bag url to transformer
> - Transformer
>   - Receive a message with the bag url
>   - Call storage service and get the bag (eg {{gateway_endpoint}}/bags/digitised/b30246039)

We decided that we needed an adapter in order to be able to do reindexes but that we would only store in a VHS the bag URL instead of the whole bag mostly because we didn't want to replicate what is already in a storage service VHS (if there was another reason to avoid storing the whole thing, I can't remember it). 
I'm not sure that we save anything in practice though, as we the METS adapter will have to store a record per bag in its VHS anyway, and since we mostly pay per operation, I don't think it will cost less than storing the whole bag.

On the other hand, having the transformer make requests to storage service, means that, every time we reindex, we will make a lot of requests to the storage service. This means:
- we might end up making too many requests to the storage service if we scale up the transformer too much and possibly need to implement some rate limiting. This seems a bit awkward as one of the purposes of having an adapter should be decoupling the pipeline from the source.
- the transformer becomes a little bit more unreliable because it might fail for reasons other than issues with the data

Basically, I'm no longer convinced that it's a good idea to have the transformer make the request to the storage service. I'm thinking that that should be the responsibility of the adapter and that it should store in its VHS the whole bag and send the whole bag to the transformer. The downside is that we end up pretty much duplicating the storage service bag VHS but I think, that's okay maybe?

**Comment by warrd** [comment]
:thinking: I can see your point.

By storing the whole bag you mean just storing the json returned from the API, rather than the METS files or anything, right?

The obvious issue is if the API reponse changes inbetween a replay, in which case this method would not capture the change, but we are not expecting that to happen I don't think?

**Comment by alicefuzier** [comment]
Yes, I meant the adapter should just store the bag JSON, not the actual files. The transformer should receive a bag JSON, find the entry for the METS file and fetch it form the storage service s3 bucket as we said before.

If the bag api response changes because of an update to a bag, the adapter would receive a notification for the update and would update the record in its VHS accordingly and then send the updated JSON to the transformer, so I don't think that's a problem. 
If it changes because there is a change in the bag api model (for example it moves from v1 to v2 and there are non backwards compatible changes) we need to change the transformer accordingly and rerun a full resync of the adapter VHS. 
So yeah, in the other design, we don't need to resync the VHS if the bag api changes model, but I think it's still ok if we achieve better decoupling?

**Comment by jamesgorrie** [comment]
Given this is the METS adapter, why not read the METS file and store that in a VHS, without any METS data.

e.g.
__Adapter:__
1. reads bag
1. gets METS file location
1. copies METS file to S3 with dynamo reference i.e. VHS

__Transformer:__
1. Gets METS file from s3
1. reads out data ... ...

**Comment by alicefuzier** [comment]
Mmh, I see. It is more intuitive because the adapter would store and send exactly what the transformer needs...
I feel that this is just adding things for the adapter to do, though, without really simplifying the transformer? Regardless of whether the METS is stored using a VHS or not, it's always going to be in s3, so the transformer will have to go fetch the METS from s3 anyway. 
Also, the VHS is basically a layer adding protection against race conditions on top of normal s3, which we don't need for the METS file, provided that we read the correct version from the storage service.
Lastly, if we store the bag instead of just the METS, we don't need a separate adapter if/when we get to processing the ALTO files, we just need another transformer that feeds off the same adapter (probably METS adapter is a bad name, maybe bag adapter would be better). Not sure if this is actually a good thing, though. Or if it matters at all at this stage

**Comment by alicefuzier** [comment]
(I think we should probably whiteboard it tomorrow, this is getting a bit complicated)

**Comment by warrd** [comment]
The downside to this is that we will need a new adaptor for every bit of data we want to pull out the storage service, whereas the original method we could have a single storage service adaptor with multiple transformers of which METS is just one (not sure that will ever be a requirement though?).

:thinking: but apart from that maybe makes more sense, maybe worth not committing to either method yet to give us some more time to think /  chat through things.

**Comment by jamesgorrie** [comment]
> The downside to this is that we will need a new adaptor for every bit of data we want to pull out the storage service

We should be trying to write single use apps most of the time anyway I suspect. Spinning up a new adapter should be easy, if not, we should be making it easy.

**Comment by warrd** [comment]
I personally don't think the advantages of multiple separate apps outweigh the disadvantages in these cases, and think proliferation of services that have very similar purposes and logic should generally be avoided. But I guess that is a broader point, and partly down to my personal preference / expertise.

**Comment by jamesgorrie** [comment]
@warrd definitely point taken - let's chat about the pros and cons when we're all in the office as you said.

**Comment by warrd** [comment]
i.e. one app that knows how  to access /  fetch data from storage-service for the pipeline seems much easier, and a better separation of concerns than having multiple with duplicated logic. Even if it is easy to spin up another adaptor, it is a much higher maintenance cost

**Comment by alicefuzier** [comment]
For me, the main difference is that having separate specific adapters allows us to deal with updates more granularly. I.e if a bag is updated but the METS file has not been changed, a METS adapter would be able to recognise that it doesn't need to forward a notification whereas a bag adapter would not be able to make a distinction.

I don't think this is actually a requirement for METS (we can just reprocess them every time) but for other things like ALTO that might be different.

If we can just reprocess every update as a new ingest, then I think I agree that I would prefer to avoid service proliferation.

**Comment by warrd** [comment]
With this method, is there any reason we have to make a copy of the file on S3? Could we not just store a hash of the file in the adapter, so we can check whether it has changed, and then emit the S3 URL to the transformer in case of a change?

**Comment by alicefuzier** [comment]
Probably no. I don't think we even need to read the actual METS in the adapter. If the METS file changes, the path to the file in the bag manifest changes because the storage service never replaces a file, it always creates new versions. So effectively it goes from `"path": "v1/data/b30246039.xml"` to `"path": "v2/data/b30246039.xml"`. In theory, we might get away with just storing the path, without actually fetching it in the adapter, and checking if that has changed before sending a notification to the transformer.
